### PR TITLE
feat: prefetch aliased fields with filters/ordering

### PR DESCRIPTION
**Pull request by rcybulski1122012**
Adjust the optimizer so it works if someone uses aliases to retrieve a field (with default resolver) few times with different argument

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Continuation of changes for https://github.com/strawberry-graphql/strawberry-django/issues/867

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

## Summary by Sourcery

Support optimized prefetching for aliased Django relation fields with differing filters and ordering in GraphQL queries.

New Features:
- Enable optimizer to prefetch aliased relation fields that use different filters or ordering by mapping each alias to a dedicated attribute.

Enhancements:
- Adjust optimizer hint generation to avoid conflicting prefetches on connection/paginated fields when alias-based prefetching is requested.

Tests:
- Add integration tests covering aliased milestones with different/same filters, ordering, and combined filters plus ordering to verify optimized query counts and results.

## Summary by Sourcery

Optimize Django relation prefetching for aliased GraphQL fields that reuse the same underlying field with different arguments.

New Features:
- Support optimized prefetching for aliased relation fields that use distinct filters or ordering by assigning each alias a dedicated to_attr-backed prefetch.

Enhancements:
- Skip optimizer-based prefetching for aliased connection or paginated fields when to_attr would conflict with their expected prefetch cache usage.
- Update field resolution to read from alias-specific prefetched attributes when no custom resolver is defined.
- Add release metadata documenting the optimizer behavior change for aliased fields with different arguments.

Tests:
- Extend optimizer integration tests to cover aliased milestones with different/same filters, ordering, and combined filters plus ordering, asserting both query counts and results.

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!

Below is the changelog that will be used for the release.

---

Adjust the optimizer to handle when the same field is selected multiple times via aliases with different arguments and the field doesn't have a custom resolver.

This release was contributed by @rcybulski1122012 in https://github.com/strawberry-graphql/strawberry-django/pull/882

File: tests/test_optimizer.py [hunk 1]
```diff
@@ -1230,6 +1230,281 @@ def test_query_prefetch_with_aliases_same_field(db, gql_client: GraphQLTestClien
     }
 
 
+@pytest.mark.django_db(transaction=True)
```
[the remark this answers is not in the document]
Comment (rcybulski1122012) [reply tests/test_optimizer.py:1233]: This is already tested in test_query_nested_connection_with_filter_and_alias

**Comment by rcybulski1122012** [comment]
@bellini666 Hey, what's about this PR? I just ran into this exact case in my app. Merging this would save me from having to monkeypatch it (or making an additional query on the frontend).

File: RELEASE.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+release type: minor
+---
+
+Adjust the optimizer to handle if the same field with different arguments using aliases is selected and the field doesn't have a custom resolver.
```
[the remark this answers is not in the document]
Comment (bellini666) [reply RELEASE.md]: I like the wording copilot suggested here

File: tests/test_optimizer.py [hunk 3]
```diff
@@ -1230,6 +1230,337 @@ def test_query_prefetch_with_aliases_same_field(db, gql_client: GraphQLTestClien
...
+        }
+      }
+    """
+
+    project_1 = ProjectFactory.create()
+    milestone_1a = MilestoneFactory.create(project=project_1, name="a")
+    project_2 = ProjectFactory.create()
+    milestone_2a = MilestoneFactory.create(project=project_2, name="a")
+
+    with assert_num_queries(3 if DjangoOptimizerExtension.enabled.get() else 5):
```
Comment (bellini666) [inline tests/test_optimizer.py]: I think if we change `_get_field_arguments` to:

```python
def _get_field_arguments(node: FieldNode) -> tuple:
    return tuple(
        print_ast(a) for a in sorted(node.arguments or (), key=lambda a: a.name.value)
    )
```

Which also drops `loc` from the node, it would reduce this to 2, which is what we want, right? Does that make sense

**Review by bellini666 — CHANGES_REQUESTED** [review]
Sorry, completely missed this 🙇‍♂️

This is great in general :)

Left a comment here which might be worth trying. Also, would it be possible to add 2 new tests? nested selections under an aliased field `(a: milestones { issues { id } })` and differing pagination args per alias. That would cover most (if not all) corner cases that are not being covered here

**Comment by rcybulski1122012** [comment]
@bellini666 Hey, I closed this PR, because I created a new one that derives from that but also handles custom resolver with parametrized optimizer hints.

https://github.com/strawberry-graphql/strawberry-django/pull/930
